Mathlib/Combinatorics/SimpleGraph/EdgeLabelling.lean

```diff
@@ -0,0 +1,217 @@
+/-
+Copyright (c) 2023 Bhavik Mehta. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Authors: Bhavik Mehta, Olivia Röhrig
+-/
+import Mathlib.Combinatorics.SimpleGraph.Maps
+import Mathlib.Data.Fintype.BigOperators
+import Mathlib.Data.ZMod.Defs
+
+/-!
+# Edge labellings
+
+This module defines labellings of the edges of a graph.
+
+## Main definitions
+
+- `SimpleGraph.EdgeLabelling`: An assignment of a label from a given type to each edge of the graph.
+
+- `SimpleGraph.EdgeLabelling.labelGraph`: the graph consisting of all edges with a given label.
+-/
+
+open Finset
+open Fintype (card)
+
+namespace SimpleGraph
+
+variable {V V' : Type*} {G : SimpleGraph V} {G' : SimpleGraph V'} {K K' : Type*}
+
+/-- An edge labelling of a simple graph `G` with labels in type `K`. Sometimes this is called an
+edge-colouring, but we reserve that terminology for labellings where incident edges cannot share a
+label.
+-/
+def EdgeLabelling (G : SimpleGraph V) (K : Type*) :=
+  G.edgeSet → K
```

Comment:
Should this be an `abbrev`?

Reply:
I also don't think so, since it's useful to treat it like a function without requiring all the FunLike machinery.

Reply:
So you want it to be reducible enough that it looks like a function, but opaque enough that rw/simp cannot rewrite its domain? This sounds like a recipe for disaster 😦